src/gt4py/backend/gt_backends.py

```diff
@@ -379,144 +465,184 @@ def visit_StencilImplementation(self, node: gt_ir.StencilImplementation):
-        "gtcache_size": {"versioning": True},
-        "debug_mode": {"versioning": True},
-        "add_profile_info": {"versioning": True},
+    @classmethod
+    @abc.abstractmethod
+    def generate_extension(
+        cls,
+        stencil_id: gt_definitions.StencilID,
+        definition_ir: gt_ir.StencilDefinition,
+        options: gt_definitions.BuildOptions,
+        **kwargs,
+    ):
+        pass
+
+
+@gt_backend.register
+class GTX86Backend(BaseGTBackend):
+
+    GT_BACKEND_T = "x86"
+
+    name = "gtx86"
+    options = BaseGTBackend.GT_BACKEND_OPTS
+    storage_info = {
+        "alignment": 1,
+        "device": "cpu",
+        "layout_map": make_x86_layout_map,
+        "is_compatible_layout": x86_is_compatible_layout,
+        "is_compatible_type": gtcpu_is_compatible_type,
     }
 
     @classmethod
-    def get_pyext_module_name(cls, stencil_id, *, qualified=False):
-        module_name = cls.get_stencil_module_name(stencil_id, qualified=qualified) + "_pyext"
-        return module_name
+    def generate_extension(cls, stencil_id, definition_ir, options, **kwargs):
+        return cls._generic_generate_extension(
+            stencil_id, definition_ir, options, uses_cuda=False, **kwargs
+        )
 
-    @classmethod
-    def get_pyext_class_name(cls, stencil_id):
-        module_name = cls.get_stencil_class_name(stencil_id) + "_pyext"
-        return module_name
 
-    @classmethod
-    def get_pyext_build_path(cls, stencil_id):
-        path = os.path.join(
-            cls.get_stencil_package_path(stencil_id),
-            cls.get_pyext_module_name(stencil_id) + "_BUILD",
-        )
+@gt_backend.register
+class GTMCBackend(BaseGTBackend):
 
-        return path
+    GT_BACKEND_T = "mc"
 
-    @classmethod
-    def generate_extension(cls, stencil_id, implementation_ir, options):
-        raise NotImplementedError(
-            "'generate_extension()' method must be implemented by subclasses"
-        )
+    name = "gtmc"
+    options = BaseGTBackend.GT_BACKEND_OPTS
+    storage_info = {
+        "alignment": 8,
+        "device": "cpu",
+        "layout_map": make_mc_layout_map,
+        "is_compatible_layout": mc_is_compatible_layout,
+        "is_compatible_type": gtcpu_is_compatible_type,
+    }
 
     @classmethod
-    def generate_cache_info(cls, stencil_id, extra_cache_info):
-        cache_info = super(BaseGTBackend, cls).generate_cache_info(stencil_id, {})
+    def generate_extension(cls, stencil_id, definition_ir, options, **kwargs):
+        return cls._generic_generate_extension(
+            stencil_id, definition_ir, options, uses_cuda=False, **kwargs
+        )
 
-        cache_info["pyext_file_path"] = extra_cache_info["pyext_file_path"]
-        cache_info["pyext_md5"] = hashlib.md5(
-            open(cache_info["pyext_file_path"], "rb").read()
-        ).hexdigest()
 
-        return cache_info
+class GTCUDAPyModuleGenerator(gt_backend.CUDAPyExtModuleGenerator):
+    def generate_pre_run(self) -> str:
```

Comment:
I like the pre-run/post-run idea much better than the backend-specific synchronize/mark modified!

Reply:
:thumbsup: